### ISSUE TITLE
Installation Problem

### ISSUE THREAD

**Issue by mrmayfield** [Acly/krita-ai-diffusion#374]
I am having problems installing the plugin. I am using Fedora 39 and Krita 5.2.2, installed as an AppImage (downloaded from the Krita website).

When I try an install the plugin, it seems to fail when trying to install mediapipe. Here is the relevant log output:

2024-02-01 08:39:29,422 ERROR Error during installation: ERROR: Could not find a version that satisfies the requirement mediapipe (from versions: none)
ERROR: No matching distribution found for mediapipe

[notice] A new release of pip is available: 23.2.1 -> 23.3.2
[notice] To update, run: python3 -m pip install --upgrade pip
Traceback (most recent call last):
  File "/home/miltonmayfield/.local/share/krita/pykrita/ai_diffusion/server.py", line 272, in install
    await self._install(cb)
  File "/home/miltonmayfield/.local/share/krita/pykrita/ai_diffusion/server.py", line 151, in _install
    await install_if_missing(dir, self._install_custom_node, pkg, network, cb)
  File "/home/miltonmayfield/.local/share/krita/pykrita/ai_diffusion/server.py", line 579, in install_if_missing
    await try_install(path, installer, *args)
  File "/home/miltonmayfield/.local/share/krita/pykrita/ai_diffusion/server.py", line 574, in try_install
    raise e
  File "/home/miltonmayfield/.local/share/krita/pykrita/ai_diffusion/server.py", line 569, in try_install
    await installer(*args)
  File "/home/miltonmayfield/.local/share/krita/pykrita/ai_diffusion/server.py", line 229, in _install_custom_node
    await _execute_process(pkg.name, self._pip_install("-r", requirements_txt), folder, cb)
  File "/home/miltonmayfield/.local/share/krita/pykrita/ai_diffusion/server.py", line 563, in _execute_process
    raise Exception(f"Error during installation: {errlog}")
Exception: Error during installation: ERROR: Could not find a version that satisfies the requirement mediapipe (from versions: none)
ERROR: No matching distribution found for mediapipe

[notice] A new release of pip is available: 23.2.1 -> 23.3.2
[notice] To update, run: python3 -m pip install --upgrade pip

2024-02-01 08:39:29,422 ERROR Installation failed

(I tried upgrading pip, but received the message that I already had version 23.3.2)

After searching a bit online, I saw that Python versions after 3.10 don't support mediapipe.

Thanks in advance for your help with this issue.

**Comment by Acly** [comment]
> After searching a bit online, I saw that Python versions after 3.10 don't support mediapipe.

It works with 3.11 (tested on Ubuntu). Which version of Python do you have? 3.12 is too new I think, also some other packages don't support it yet

**Comment by Acly** [comment]
You'll have to install python3.11 package and make it the default python3.
Are you using python for anything else where this might interfere?

Instead of changing the default it might work to modify [server.py line 83](https://github.com/Acly/krita-ai-diffusion/blob/main/ai_diffusion/server.py#L83) and change `python3` to `python3.11`

**Comment by mrmayfield** [comment]
Thanks for all of your help. Using python 3.11 fixed the problem. I installed PyEnv to allow easy switching between my different python versions, and the plugin installed fine.

I appreciate your timely help and everything you have done with this plugin.

**Comment by SimonHeggie** [comment]
Acly,

Regarding this issue, If I'm not mistake this seems like Pytorch not providing adequate API handles for required versions of dependencies.

It's not particularly elegant but perhaps for the sake of the health of support on this project, there is something we can do on our end here.

Looking at something like ChaiNNer; can we use scraping to inform the server install process where to go to get a compatible version of python for Pytorch as ChaiNNer does? Fall back of having more than one method similar to mirror downloading. Either way, including a fallback method of a hardcoded '3.11' will retain the kinds of people who'd usually misunderstand and therefore reject AI technology and walk away in frustration.

I'm seriously torn as to whether I try to patch my own install script to do this in terminal. I imagine the problem goes away if PyTorch begin including pyton dependency metadata somewhere.

Like:

```
import asyncio
import requests

async def get_compatible_python_version():
    # Attempt 1: Fetch from PyPI
    try:
        response = requests.get('https://pypi.org/pypi/torch/json')
        data = response.json()
        requires_python = data['info']['requires_python']
        if requires_python:
            return requires_python.split(',')[0].strip()  # Simplified assumption
    except Exception as e:
        print(f"Error fetching from PyPI: {e}")

    # Fallback: Hardcoded version
    return "3.11"

async def setup_python_environment(python_version):
    # Check if pyenv is installed; if not, install it
    if not shutil.which("pyenv"):
        # Pseudo-code for installing pyenv
        await execute_command(["curl https://pyenv.run | bash"])
        # Add pyenv init to shell (this will vary based on the shell)

    # Install required Python version using pyenv
    await execute_command(["pyenv", "install", python_version])

    # Set the local (project-specific) Python version
    await execute_command(["pyenv", "local", python_version])

```

Probably in practice would look like:

`class Server:
    # Existing attributes...
    _python_cmd: Optional[Path] = None
    _pip_cmd: Optional[Path] = None
    _cache_dir: Path
    _version_file: Path
    _process: Optional[asyncio.subprocess.Process] = None
    _task: Optional[asyncio.Task] = None

    def __init__(self, path: Optional[str] = None):
        # Existing initialization code...
        self.check_install()

    async def check_python_version_and_setup(self):
        compatible_version = await get_compatible_python_version()
        # Assuming `compatible_version` returns a simple version string like '3.8'
        # You might need to parse or choose a specific version based on the compatibility range returned
        await self.setup_python_environment(compatible_version)

    async def setup_python_environment(self, python_version):
        # Check if pyenv is installed; if not, install it
        if not shutil.which("pyenv"):
            log.info("Installing pyenv...")
            # Installation command may vary depending on the OS
            install_cmd = "curl https://pyenv.run | bash"
            subprocess.run(install_cmd, shell=True, check=True)
            # Add pyenv init to shell is skipped for simplicity; typically involves modifying shell profile scripts

        # Install and set the required Python version using pyenv
        log.info(f"Installing Python {python_version} using pyenv...")
        subprocess.run(f"pyenv install -s {python_version}", shell=True, check=True)
        subprocess.run(f"pyenv local {python_version}", shell=True, check=True)
        self._python_cmd = Path(f"$(pyenv which python)")
        self._pip_cmd = Path(f"$(pyenv which pip)")

    def check_install(self):
        # Existing check_install logic...
        # After checking comfy and python installations, add:
        asyncio.run(self.check_python_version_and_setup())

    # Existing methods...
    async def _install(self, cb: InternalCB):
        # Adjust the existing installation process to consider the new Python environment setup
        # This might involve skipping the Python setup if it's already done by `check_python_version_and_setup`

    # Rest of the Server class...

# Additional utility and helper functions as before...
`

Thoughts?

-S

**Comment by Acly** [comment]
* I think adding `python3.11` to the start of the search path is a good idea, will do this later
* PyTorch now released version 2.2 which support Python 3.12 actually, but other dependencies (like the one this issue is about) do not
* Including PyEnv in the mix doesn't help that much, it won't work if people are missing packages like `python-dev`. So either way they have to use package manager to install *something*, might as well go for `python3.11` directly

**Comment by SimonHeggie** [comment]
Sounds like a plan.

I generally don't know what dependencies are needed for most software I use. Even my favourite FOSS tool Blender, I constantly forget the python version being used.
I'm not even confident with PyEnv having never personally handled it myself.

But the majority of artists I've worked are in that boat or even more oblivious. Even some of the technical artists!

I might just wait for the fix before I jump back into testing with my laptop RTX 4090.

**Comment by Acly** [comment]
Compared to other software the ecosystem for SD/GenAI is just very experimental and unstable still. The installer is a best effort to automate things, and I appreciate efforts to improve it, But there is only so much you can do while also staying flexible to follow latest developments. IMO a one-click it-just-works solution for everyone is currently only possible via cloud service (which would also work for the 99% who don't have a 4090 laptop, but has other downsides of course).

**Comment by SimonHeggie** [comment]
I agree that it's not wise to build complex code structures around the shifting sands of unstable development.
However, I thought we were discussing the simpler fix of changing a line to hard-code it to use python 3.11?

Unfortunately the server route is outside my option set. I'm still paying off the laptop. Typical artist existence.

Now as to whether you should place any kind of priority into ensuring artists, particularly Krita users, can use your tool?
This get's in a very sensitive topic which I will continue in PM with you.

Please check your PM's in Krita Aritsts.

**Comment by SimonHeggie** [comment]
That last commit works when I copy and past that onto line 83 of the server.py in the addon directory.

![image](https://github.com/Acly/krita-ai-diffusion/assets/86150800/d54b8f15-8ff1-473e-91b7-d421ab3e3233)

ITS BEAUTIFUL.

**Comment by pflovv** [comment]
Previously, I tried to downgrade the pacman Python, although Krita worked, it destroyed my caffeine-ng and redshift. It turned out I couldn't stick with only one python version. After searching around, I realized there's no need to rollback to Python 3.11, as different version of Python can coexist as it creates isolated virtual environments for different usage. For example, I use Arch, and I installed python311 package from AUR alongside my pacman Python3.13, everything worked now.
